osu.Game/Database/RealmContextFactory.cs

```diff
@@ -61,33 +63,60 @@ public class RealmContextFactory : IDisposable
 
         private readonly ThreadLocal<bool> currentThreadCanCreateContexts = new ThreadLocal<bool>();
 
+        /// <summary>
+        /// Holds a map of functions registered via <see cref="RegisterCustomSubscription"/> and <see cref="RegisterForNotifications{T}"/> and a coinciding action which when triggered,
+        /// will unregister the subscription from realm.
+        ///
+        /// Put another way, the key is an action which registers the subscription with realm. The returned <see cref="IDisposable"/> from the action is stored as the value and only
+        /// used internally.
+        ///
+        /// Entries in this dictionary are only removed when a consumer signals that the subscription should be permanently ceased (via their own <see cref="IDisposable"/>).
+        /// </summary>
+        private readonly Dictionary<Func<Realm, IDisposable?>, IDisposable?> customSubscriptionsResetMap = new Dictionary<Func<Realm, IDisposable?>, IDisposable?>();
+
+        /// <summary>
+        /// Holds a map of functions registered via <see cref="RegisterForNotifications{T}"/> and a coinciding action which when triggered,
+        /// fires a change set event with an empty collection. This is used to inform subscribers when a realm context goes away, and ensure they don't use invalidated
+        /// managed realm objects from a previous firing.
+        /// </summary>
+        private readonly Dictionary<Func<Realm, IDisposable?>, Action> realmSubscriptionsResetMap = new Dictionary<Func<Realm, IDisposable?>, Action>();
```

Comment:
The naming of `customSubscriptionResetMap` and `realmSubscriptionResetMap` is super confusing to me, especially because they're used in different ways - `realmSubscriptionResetMap` is used when a context is lost and `customSubscriptionResetMap` is used when a new context is created.

Perhaps the former should be named `subscriptionMap` and the latter be called `subscriptionResetMap`?

Reply:
Sounds good. Naming is hard.

Reply:
Actually that's not the case. Both maps are used on context loss.

Reply:
Right, `realmSubscriptionResetMap` -> `notificationResetMap` then?

Reply:
Went with `notificationsResetMap` to match the method name